teps/0007-conditions-beta.md

```diff
@@ -0,0 +1,610 @@
+      fi
+```
+
+### Efficiency
+To improve the efficiency of guarded execution of `Tasks`, we need to avoid spinning up new `Pods` to check `Guards`. We can use string expressions for `Guard` checking, but we want to avoid adding an opinionated and complex expression language to the Tekton API to ensure Tekton can be supported by as many systems as possible. We propose using a simple expression syntax `When Expressions` (similar to Kubernetes' `When Expressions`) to evaluate `Guards` efficiently. The components of `When Expressions` are `Input`, `Operator` and `Values`. `Input` is the input for the `Guard` checking which can be static inputs or outputs from parent `Tasks`, such as `Parameters` or `Results`. `Values` is an array of string values. `Operator` represents an `Input`'s relationship to a set of `Values`. `Operators` we will use in `Guards` are `In` and `NotIn`. The `Values` array must be non-empty. When we have more than one `Guard`, the guarded `Task` will be executed when all the `Guards` evaluate to `True`. Note that when a `Guard` uses a resource from another `Task`, such as a `Result`, it introduces an implicit resource dependency that makes the guarded `Task` dependent on the resource-producing `Task`. 
+
+Here's how a user can specify a `Guard` that executes on that `Result`: 
+
+```yaml
+apiVersion: tekton.dev/v1alpha1
+kind: Task
+metadata:
+  name: file-exists
+spec:
+  params:
+    - name: path
+  workspaces:
+    - name: source
+  results:
+    - name: exists
+      description: boolean indicating whether the file exists
+  steps:
+  - name: check-file-exists
+    image: alpine
+    script: |
+      if [ -f $(workspaces.source.path)/$(params.path) ]; then
+        echo true | tee /tekton/results/exists
+      else
+        echo false | tee /tekton/results/exists
+      fi
+---
+api: tekton.dev/v1alpha1
+kind: Pipeline
+metadata: 
+    name: generate-file
+spec:
+  workspaces:
+    - name: source-repo
+  params:
+    - name: path
+      default: 'README.md'
+  tasks: 
+    - name: create-file
+      taskRef: 
+        name: create-readme-file
+      workspaces:
+        - name: source
+          workspace: source-repo
+    - name: file-exists
+      taskRef: 
+        name: file-exists
+      workspaces:
+        - name: source
+          workspace: source-repo
+    - name: echo-file-exists 
+      when:
+        - name: check-file-exists
+          input: '$(tasks.file-exists.results.exists)'
+          operator: In
+          values: ['true']
+      taskRef:
+        name: echo-file-exists
+```
+
+Other examples of `Guards` using `When Expressions` are:
+
+```yaml
+name: branch-is-main
+input: $(params.branch)
+operator: In
+values: [‘main’]
+---
+name: branch-not-main
+input: $(params.branch)
+operator: NotIn
+values: [‘main’]
+---
+name: always-false
+input: ‘false’
+operator: In
+values: [‘’]
+```
+
+We can explore adding more `Operators` later if needed, such as `IsTrue`, `IsFalse`, `IsEmpty` and `IsNotEmpty`.
```

Comment:
links to the k8s equivalents of for all of these will help - so folks know why we are mentioning these specific values and not others

(and maybe explicitly mention the fact that < and > are not included here, tho they exist in k8s match expressions?)

Reply:
added the links and clarifications, thanks